idea.md

```diff
@@ -0,0 +1,164 @@
+const total_released_tokens = new Map([['TOKEN', 0]]);
+const released_tokens_of_user = new Map([['TOKEN', new Map([['USER', 0]])]]);
++ const last_balance_of_user_for_tokens = new Map([['TOKEN', new Map([['USER', 0]])]]);
+
+function withdraw(token, user) {
+    const Property = ERC20(property);
+    const Token = ERC20(token);
+    const total_received = 
+          Token.balanceOf(address(this)) +
+          total_released_tokens.get(token);
+
++   updateReleasedTokens(token, user, Property.balanceOf(user));
+    
+    /*
+     */
+    const released = released_tokens_of_user.get(token).get(user);
+
+    const user_balance = Property.balanceOf(user);
+    const total_supply = Property.totalSupply();
+    const payment = 
+        (total_received * user_balance) / total_supply - released;
+
+    /*
+     * (TRANSFER TOKENS TO THE USER HERE...)
+     */
+    ERC20.safeTransfer(token, user, payment);
+
+    /*
+     * Update the global states.
+     */
+    released_tokens_of_user.get(token).set(user, released + payment)
+    total_released_tokens.set(
+        token,
+        total_released_tokens.get(token) + payment
+    );
+    
++   last_balance_of_user_for_tokens[token][user] = user_balance;
+
+    return payment;
+}
+```
+
+`updateReleasedTokens` updates the value of `released` by referencing `Withdraw.transferHistory` recursively until `last_balance_of_user_for_tokens` matches the latest balance.
+
+```javascript!
+const released_tokens_addresses = new Map([['USER', new Set()]]); // mapping(address => EnumerableSet.AddressSet)
+const Withdraw = Withdraw(Registry(registry).registries('Withdraw'));
+
+function updateReleasedTokens(token, user, current_balance) {
+    if (last_balance_of_user_for_tokens[token][user] === current_balance) {
+        // No need to update it.
+        return
+    }
+
+    const historyIndex = Withdraw.transferHistoryLengthOfRecipient(address(Property), user);
+    let done = false;
+    let i = historyIndex;
+    let calculated = 0;
+    while (!done || i >= 0) {
+        const history = Withdraw.transferHistory(
+            Withdraw.transferHistoryOfRecipientByIndex(i)
+        );
+        if (!history.fill) {
+            // This is always done on the history of the last 1 transfer, and other case, history.fill is true.
+            history.amount = current_balance - history.preBalanceOfRecipient;
+        }
+
+        updateReleasedTokens(token, history.from, Property.balanceOf(history.from));
+
+        const released_tokens = released_tokens_of_user.get(token).get(history.from);
+        const part_of_released = released_tokens * history.preBalanceOfSender / history.amount;
+        released_tokens_of_user.get(token).set(
+            user,
+            part_of_released + released_tokens_of_user.get(token).get(user)
+        );
+        
+        i = i - 1;
+        calculated = calculated + history.amount;
+        done = last_balance_of_user_for_tokens[token][user] + calculated === current_balance;
```

Comment:
If the calculated transfers + recorded balance matches the user's latest balance, the loop ends.